File: go/redblacktree/red_black_tree.go
package redblacktree

const (
	RED = true
	BLACK = false
)

type Node struct {
	key string
	value int
	left *Node
	right *Node
	// 具体 true 为红还是黑，自己定
	color bool
}

func NewNode(key string, value int) *Node {
	// 在 2-3 树中添加一个节点，肯定是要进行融合节点，因此新创建的节点我们都设置为红节点
	// 代表这是一个和其父节点融合的节点
	return &Node{
		key: key,
		value: value,
		color: RED,
	}
}

type RBTree struct {
	root *Node
	size int
}

func (this *RBTree) IsRed(node *Node) bool {
	if node == nil {
		return BLACK
	}
	return node.color
}

func (this *RBTree) Add(key string, value int) {
	this.root = this.add(this.root, key, value)
	this.root.color = BLACK
}


func (this *RBTree) add(root *Node, key string, value int) *Node {
	if root == nil {
		node := NewNode(key, value)
		this.size++
		return node
	}

	if key < root.key {
		root.left = this.add(root.left, key, value)
	} else if key > root.key {
		root.right = this.add(root.right, key, value)
	} else {
		root.value = value
	}

	if this.IsRed(root.right) && !this.IsRed(root.left) {
		root = this.leftRotate(root)
	}

	if this.IsRed(root.left) && this.IsRed(root.left.left) {
		root = this.rightRotate(root)
	}

	if this.IsRed(root.left) && this.IsRed(root.right) {
		this.filpColors(root)
	}

	return root
}

// 首先，在一个 2 节点中添加一个节点，必然要进行融合
// 如何新节点融合到左边，那没什么问题，添加为新子节点就好
// 但是如果融合到右边，那要进行一次左旋过程，并且旋转后的左节点要变红
func (this *RBTree) leftRotate(root *Node) *Node {
	x := root.right
	root.right = x.left
	x.left = root
	x.color = root.color
	root.color = RED
	return x
}

// 当对一个 3 节点中添加一个右节点时，要进行颜色反转
func (this *RBTree) filpColors(node *Node) {
	node.color = RED
	node.left.color = BLACK
	node.right.color = RED 
}

// 当对一个 3 节点中添加一个左节点时，要进行右旋转
func (this *RBTree) rightRotate(node *Node) *Node {
	x := node.left
	node.left = x.right
	x.right = node
	x.color = node.color
	node.color = RED
	return x
}

File: js/redBlackTree/redBlackTree.js
class Node {
  constructor(key, value) {
    this.key = key;
    this.value = value;
    this.red = true;
    this.left = null;
    this.right = null;
  }
}

class RBTree {
  constructor() {
    this.root = new Node(null, null);
    this.root.red = false;
    this.size = 0;
  }
  

  getSize() {
    return this.size;
  }

  isEmpty() {
    return this.size === 0;
  }

  insert(key, value) {
    this.root = this.insertInTree(this.root, key, value);
    this.root.red = false;
  }

  insertInTree(node, key, value) {
    if (node === null) {
      const newNode = new Node(key, value);
      this.size++;
      return newNode;
    }

    if (key < node.key) {
      node.left = this.insertInTree(node.left, key, value);
    }

    if (key > node.key) {
      node.right = this.insertInTree(node.right, key, value);
    }

    if (key === node.key) {
      node.value = value;
    }

    // 回想我们的整个流程
    // 完整的应该是 左旋 -> 右旋 -> 反转

    // 如果右红左不红，2- 节点右插入和 3- 节点右插入
    if (this.isRed(node.right) && !this.isRed(node.left)) {
      node = this.rotateLeft(node);
    }

    // 左左双红
    if (this.isRed(node.left) && this.isRed(node.left.left)) {
      node = this.rotateRight(node);
    }

    if (this.isRed(node.left) && this.isRed(node.right)) {
      this.flipColor(node);
    }
  }

  isRed(node) {
    if (node === null) {
      return false;
    }
    return node.red;
  }

  rotateLeft(node) {
    const x = node.right;
    node.right = x.left;
    x.left = node;
    x.red = node.red;
    node.red = true;
    return x;
  }

  rotateRight(node) {
    const x = node.left;
    node.left = x.right;
    x.right = node;
    x.color = node.color;
    node.red = true;
  }

  flipColor(node) {
    // 这里不需要添加 node.left 和 node.right 的判断
    // 为什么呢，因为我们仔细想想什么时候会需要颜色反转呢
    // 3- 节点的右插入，那必然左右节点都不为空
    node.left.red = false;
    node.right.red = false;
    node.red = true;
  }


}